### DISCUSSION TITLE
Need to manage in app purchase when app is in background

### DISCUSSION THREAD

**Discussion by Sneha1231** [hyochan/react-native-iap#1349] [Q&A]
Hi,

I initated an in app purchase from my app, then I force quit the app. After that I got in app purchase prompt and able to do the purchase at apple side. Since I have force quoted the app I missed the server updation and this trasaction won't be saved any where.

How to manage this situation.
Is there a way to identify the app is in foreground and disable purchsae button in In app purchase prompt if not.

**Comment by neilco** [comment]
You haven’t mentioned which platform you’re referring to and the answer is different for Android than for iOS. I’ll only deal with iOS here and let someone with a greater insight address the Android side. 

If you force quit your app immediately after initiating the App Store purchase flow and then complete the purchase, the transaction will still be in StoreKit payment queue on your device in an unfinished state. This mean the next time you launch your app, you can fetch the pending purchases from the payment queue and send the receipts to your backend. The trick, as pointed out in the Apple documentation, is to set up your delegates as early as possible in the app lifecycle. From a React Native angle, this means subscribing to the `purchaseUpdatedListener` early. Just remember to call `finishTransaction` after uploading each transaction you receive to your server and the pending transactions will be cleared from the payment queue.  

One benefit of setting up this listener early (and keeping it around for the full lifecycle) this is that it will also capture transactions that occur _outside_ of your app, e.g. subscription automatic renewals, purchases made directly in the App Store, etc. It just requires your customers to open the app every now and then.

**Reply by Sneha1231** [reply]
Thanks for the reply. The platform is iOS. I just figured out that if we force quit the app after initiating purchase, in app purchase won't let us to make payment. Actually it's showing invalid purchase in in app popup. This behaviour is only available in production and won't be there in sandbox. Apple is internally handling in in SKPayment Iguess.
